easy_transformer/hook_points.py

```diff
@@ -1,149 +1,216 @@
+        del self.context
+        self.context = {}
 
     def forward(self, x):
+        """Identity function"""
         return x
 
     def layer(self):
-        # Returns the layer index if the name has the form 'blocks.{layer}.{...}'
-        # Helper function that's mainly useful on EasyTransformer
-        # If it doesn't have this form, raises an error -
-        split_name = self.name.split(".")
-        return int(split_name[1])
+        """
+        Returns the layer index if the name has the form 'blocks.{layer}.{...}'
+        Helper function that's mainly useful on EasyTransformer
+        If it doesn't have this form, raises an error -
+        """
+        _, layer, _ = self.name.split(".")
+        return int(layer)
 
 
 # %%
+# TODO maybe this should be a decorator so clients don't have to call setup()?
+# or just a class should be enough?
 class HookedRootModule(nn.Module):
     """
     A class building on nn.Module to interface nicely with HookPoints
     Adds various nice utilities, most notably run_with_hooks to run the model with temporary hooks, and run_with_cache to run the model on some input and return a cache of all activations
 
+    See [EasyTransformer_Demo.ipynb] for an example of how to use it.
+
     WARNING: The main footgun with PyTorch hooking is that hooks are GLOBAL state. If you add a hook to the module, and then run it a bunch of times, the hooks persist. If you debug a broken hook and add the fixed version, the broken one is still there. To solve this, run_with_hooks will remove hooks at the start and end by default, and I recommend using reset_hooks liberally in your code.
 
     The main time this goes wrong is when you want to use backward hooks (to cache or intervene on gradients). In this case, you need to keep the hooks around as global state until you've run loss.backward() (and so need to disable the reset_hooks_end flag on run_with_hooks)
     """
 
-    def __init__(self, *args):
+    def __init__(self):
         super().__init__()
-        self.is_caching = False
+        self.is_caching = False  # TODO what's this used for?
 
     def setup(self):
-        # Setup function - this needs to be run in __init__ AFTER defining all
-        # layers
-        # Add a parameter to each module giving its name
+        """
+        Setup function - this needs to be run in the model's [__init__]
+        AFTER defining all layers.
+        """
+        # Add a parameter to each module giving its name.
+        self.name_to_module = {}
         # Build a dictionary mapping a module name to the module
-        self.mod_dict = {}
-        self.hook_dict = {}
+        self.name_to_hook = {}
         for name, module in self.named_modules():
             module.name = name
-            self.mod_dict[name] = module
+            self.name_to_module[
+                name
+            ] = module  # TODO why not just use self.named_modules()?
+            # TODO is this really the best way to do this?
```

Comment:
Yes. It's such a mess

Reply:
I tried man. I tried.

Reply:
what about a subclass?

Reply:
(and then checking https://www.programiz.com/python-programming/methods/built-in/issubclass)